reps/2023-06-06-simplify-sync.md

```diff
@@ -0,0 +1,127 @@
+
+@pcmoritz
+
+## User Story
+
+Persistence will center around the recently introduced `storage_path` option.
+
+### Persistent trial directory
+
+Ray Train provides the abstraction of a persistent trial directory. Each trial has its current working directory set to a trial specific directory (e.g., `~/ray_results/experiment_1/trial_1`). Train will synchronize this trial directory with persistent storage specified via the `storage_path` option.
+
+This directory is persistent since Train is syncing it to persistent storage for the user. The experiment driver will record experiment results and other metadata to a similar directory on the head node, which will also synchronize to persistent storage.
+
+![overview](user-story.png)
+
+### Checkpoints
+
+While trials can write arbitrary data to the persistent trial directory, they cannot be guaranteed they can read it back on checkpoint/restore. Data that is needed for resume must be recorded via the `train.report(checkpoint)` API.
+
+For example, this is how you could record a Torch checkpoint:
+
+```python
+   torch.save(model, "path/to/dir")
+   train.report(checkpoint=Checkpoint.from_directory("path/to/dir"))
+```
+
+
+Train will take ownership of the specified checkpoint data, deleting the directory after the checkpoint upload finishes. The checkpoint data will then be managed by Train (e.g., Train may restore only the latest checkpoint or delete previous checkpoints to save space according to checkpoint policy).
+
+The files of the recorded checkpoint can be accessed via `result.checkpoint`. The `Checkpoint` object itself is a logical tuple of `(path, pyarrow.fs.FileSystem)`, and this tuple is immutable. Users can also get and set arbitrary metadata to these checkpoints (e.g., preprocessor configs, model settings, etc), which will be recorded in a `metadata.json` file.
+
+### Multi-rank checkpoints
+
+Checkpoints can be recorded from multiple ranks. By default, only checkpoint data from rank zero is preserved. Data from all ranks can be retained via a `keep_all_ranks` option. Train will merge checkpoint data from all ranks into a single directory for the checkpoint number. The reason is this: users can always namespace their files/dirs by rank by the `train.context.world_rank` manually, so there isn't any reason why Train should do it for them. In fact, certain frameworks such as Lightning assume merged directories and could break if we enforced per-rank subdirectories.
+
+Similarly, for restoration the new Checkpoint API allows users to customize restoration strategy. For example, if they wanted to implement per-rank restore, they could call `pyarrow.fs.copy_files()` on selected files within the checkpoint path, rather than using the naive `Checkpoint.to_directory()` call.
```

Comment:
How about adding args to to_directory such as `include_paths` whitelist of files/dirs to include?

I was thinking it's much easier than the example you wrote though, for ex:

```
checkpoint = session.get_checkpoint()
os.makedirs("my_rank2_data")
pyarrow.fs.copy_files(checkpoint.path + "/rank_2", "my_rank2_data", source_filesystem=checkpoint.fs)
```

Reply:
I'm thinking about the deepspeed checkpoint structure, where all shards are aggregated in a single directory rather than in separate rank folders. Uploading shouldn't require any special logic if rank-specific files are named properly. Downloading may be a pain for users to interact with this low level `pyarrow.fs` interface, if they only want certain files.

I think I'm fine just sticking with the simplest option for now, which is to not add any helpers on top of `pyarrow.fs`. However, I think it'd be helpful to include some snippets that show common custom usage patterns for uploading and downloading.

Reply:
For example, how does a user do a custom upload? Do they use `pyarrow.fs.copy_files` directly, then call `session.report(..., checkpoint=Checkpoint.from_directory("s3://place/where/I/uploaded")`? In that case, would `Checkpoint.from_path` make more sense as the path of an arbitrary filesystem?

Reply:
The contract for `report` is that they provide a valid Checkpoint, which can point locally or anywhere. You can construct checkpoints pointing to arbitrary files using the raw Checkpoint constructor (e.g., `Checkpoint("s3://foo/bar")` or `Checkpoint("s3://foo/bar", filesystem)`.

Reply:
Just to walk through a concrete example of saving to different rank checkpoint folders:

```python
checkpoint_dir = "/tmp/ckpt/rank_k/model.pt"
session.report(Checkpoint("/tmp/ckpt")) # or Checkpoint("/tmp/ckpt/rank_k") with Matt's suggestion

# This will now show up as:
# storage_path/exp_dir/trial_dir/checkpoint_000x/rank_k/...
```

Is `Checkpoint.from_directory` even needed anymore? What about just having one way to create a checkpoint -- with the constructor.

Reply:
One consideration for custom uploading:

```python
for step in range(100):
    # Upload checkpoint to s3 manually
    # Ex: I want all my model checkpoints in the same bucket
    checkpoint_uri = f"s3://bucket/checkpoints/ckpt_0000{step}/"
    pyarrow.fs.copy_files("/tmp/ckpt", checkpoint_uri)

    # Report this s3 checkpoint to AIR
    session.report(Checkpoint(checkpoint_uri))

    # Which behavior makes sense here?
    # 1. Move the checkpoint to the trial directory (in which case there was no point doing this custom uploading)
    # 2. Copy the checkpoint to the trial dir
    # 3. No move/copy, just keep the checkpoint where it already is, if the user passed in a non-local URI.
```

I think we can just stick with the "move" policy as the default, and add customization options to be able to achieve any of the 3 behaviors listed above. One possibility is allowing customization of a checkpoint upload function:

```python
def default_upload_fn(checkpoint, storage):
    pyarrow.fs.copy_files(...)
    pyarrow.fs.delete(...)
    return new_checkpoint_path

def copy_upload_fn(checkpoint, storage):
    pyarrow.fs.copy_files(...)
    return new_checkpoint_path

def noop_upload_fn(checkpoint, storage):
    return checkpoint.path
```

Reply:
Actually, I think custom upload is not a very big issue. Reporting a generic local directory is already very flexible and covers most use cases.